akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringEvents.scala

```diff
@@ -0,0 +1,222 @@
+        cancelRecoveryTimer()
+        setLastSequenceNr(highestSeqNr)
+
+        try onRecoveryCompleted(state)
+        catch { case NonFatal(ex) ⇒ onRecoveryFailure(ex, Some(state)) }
+
+      case ReplayMessagesFailure(cause) ⇒
+        onRecoveryFailure(cause, event = None)
+
+      case other ⇒
+        stash(other)
+        Behaviors.same
+    }
+  } catch {
+    case NonFatal(e) ⇒
+      cancelRecoveryTimer()
+      onRecoveryFailure(e, None)
+  }
+
+  def onSnapshotterResponse(response: SnapshotProtocol.Response): Behavior[Any] = {
+    log.warning("Unexpected [{}] from SnapshotStore, already in recovering events state.", Logging.simpleName(response))
+    Behaviors.same // ignore the response
+  }
+
+  /**
+   * Called whenever a message replay fails. By default it logs the error.
+   *
+   * The actor is always stopped after this method has been invoked.
+   *
+   * @param cause failure cause.
+   * @param event the event that was processed in `receiveRecover`, if the exception was thrown there
+   */
+  protected def onRecoveryFailure(cause: Throwable, event: Option[Any]): Behavior[Any] = {
+    returnRecoveryPermit("on recovery failure: " + cause.getMessage)
+    cancelRecoveryTimer()
+
+    event match {
+      case Some(evt) ⇒
+        log.error(cause, "Exception in receiveRecover when replaying event type [{}] with sequence number [{}].", evt.getClass.getName, sequenceNr)
+        Behaviors.stopped
+
+      case None ⇒
+        log.error(cause, "Persistence failure when replaying events.  Last known sequence number [{}]", persistenceId, sequenceNr)
+        Behaviors.stopped
+    }
+  }
+
+  protected def onRecoveryCompleted(state: State): Behavior[Any] = {
+    try {
+      returnRecoveryPermit("recovery completed successfully")
+      callbacks.recoveryCompleted(commandContext, state)
+
+      val running = new EventsourcedRunning[Command, Event, State](
+        persistenceId,
+        context,
+        timers,
+        internalStash,
+
+        sequenceNr,
+        writerIdentity,
+
+        state,
+
+        callbacks,
+        pluginIds
+      )
+
+      tryUnstash(context, running)
+    } finally {
+      cancelRecoveryTimer()
+    }
+  }
+
+  protected def onRecoveryTick(snapshot: Boolean): Behavior[Any] =
+    if (!snapshot) {
+      if (!eventSeenInInterval) {
+        cancelRecoveryTimer()
+        val msg = s"Recovery timed out, didn't get event within $timeout, highest sequence number seen $sequenceNr"
+        onRecoveryFailure(new RecoveryTimedOut(msg), event = None) // TODO allow users to hook into this?
+      } else {
+        eventSeenInInterval = false
+        same
+      }
+    } else {
+      // snapshot timeout, but we're already in the events recovery phase
+      Behavior.unhandled
+    }
+
+  // ----------
+
+  override def onMessage(msg: Any): Behavior[Any] = {
+    msg match {
+      // TODO explore crazy hashcode hack to make this match quicker...?
```

Comment:
Is it slow?

Reply:
Perhaps not impactful enough as it will be slow due to database anyway; "normal operation" is missing all 3 cases, and entering last, so I thought to work around that, but maybe not indeed